File: AccountTransferMvpApp/AccountTransferMvpAppTests/Presenters.Tests.Unit/AccountTransferPresenterTests.m
#import "AccountTransferPresenterTests.h"
#import "OCMockObject.h"
#import "IAccountTransferView.h"
#import "IRemoteAccountRepository.h"
#import "ILocalAccountRepository.h"
#import "OCMockRecorder.h"
#import "AccountTransferPresenter.h"

@implementation AccountTransferPresenterTests

-(void)test_given_sut_when_I_transfer_amount_then_all_delegated_repositories_are_called {
    id accountTransferView = [OCMockObject mockForProtocol:@protocol(IAccountTransferView)];
    id remoteAccountRepository = [OCMockObject mockForProtocol:@protocol(IRemoteAccountRepository)];
    id localAccountRepository = [OCMockObject mockForProtocol:@protocol(ILocalAccountRepository)];
    
    // expectations
    NSNumber *amount = @150.0;
    [[[accountTransferView expect] andReturn:amount] getTransferAmount];
    [[remoteAccountRepository expect] withdrawAmount:amount];
    [[localAccountRepository expect] depositAmount:amount];
    [[accountTransferView expect] setDisplayMessage:@"$150 transferred."];

    AccountTransferPresenter *sut = [[AccountTransferPresenter alloc] initWithRemote:remoteAccountRepository andLocal:localAccountRepository];
    sut.accountTransferView = (id<IAccountTransferView>)accountTransferView;
    [sut transferAmount];

    [accountTransferView verify];
    [remoteAccountRepository verify];
    [localAccountRepository verify];
}

@end

File: AccountTransferMvpApp/UnitAndIntegrationTests/src/com/gaddzeit/presenters/tests/unit/AccountTransferPresenterTests.java
package com.gaddzeit.presenters.tests.unit;

import com.gaddzeit.presenters.AccountTransferPresenter;
import com.gaddzeit.repositories.ILocalAccountRepository;
import com.gaddzeit.repositories.IRemoteAccountRepository;
import com.gaddzeit.views.IAccountTransferView;
import org.jmock.Expectations;
import org.jmock.Mockery;
import org.junit.Test;

public class AccountTransferPresenterTests {
    @Test
    public void givenSut_whenTransferAmountRequested_thenAllDelegatedInterfacesAreCalled() {
        final Mockery mockingContext = new Mockery();
        final IAccountTransferView accountTransferView = mockingContext.mock(IAccountTransferView.class);
        final IRemoteAccountRepository remoteAccountRepository = mockingContext.mock(IRemoteAccountRepository.class);
        final ILocalAccountRepository localAccountRepository = mockingContext.mock(ILocalAccountRepository.class);

        // set expectations
        final double amount = 150.0;
        mockingContext.checking(new Expectations() {
            {
                one(accountTransferView).getTransferAmount();
                will(returnValue(amount));
                one(remoteAccountRepository).withdrawAmount(amount);
                one(localAccountRepository).depositAmount(amount);
                one(accountTransferView).setDisplayMessage("$150.0 transferred.");
            }
        });

        // instantiate sut
        final AccountTransferPresenter sut = new AccountTransferPresenter(remoteAccountRepository, localAccountRepository);
        sut.setView(accountTransferView);
        sut.transferAmount();

        // verify
        mockingContext.assertIsSatisfied();
    }
}
